### ISSUE TITLE
[python] Develop a consistent rule for what goes into the top-level namespace

### ISSUE THREAD

**Issue by johnkerl** [single-cell-data/TileDB-SOMA#915]

**Comment by thetorpedodog** [comment]
My thought has been "things we expect a user to instantiate themselves or call directly". So the basic SOMA types, globals like `open`, and enums like `somacore.ResultOrder` belong there. If it's something we only ever expect to *return* from a call (like the `SparseRead` implementation), that does *not* need to go in the root namespace.

**Comment by johnkerl** [comment]
Throwing out another idea:

* Anything listed _in the spec_ goes at top level
* Anything else (whatsoever) goes into a submodule -- harsh, I know, but definitely a crisp and _followable_ rule without subjectivity

**Comment by johnkerl** [comment]
@thetorpedodog I like your suggestion a lot! Although, that logic would have us include `soma.io` at top level (it is solidly user-facing) & I think we want that in a module ... 🤔

**Comment by bkmartinjr** [comment]
> things we expect a user to instantiate themselves or call directly

This makes sense to me.  The only place I am waffling is storage-system-specific names, which I want to clearly disambiguate from other storage-system-specific names, for the day when we have multiple implementations.  The alternative is that we could require these names to clearly disambiguate in their own name (eg, 'soma.SOMATileDBContext` vs `soma.tiledb.SOMAContext`).  Either would work for me....

We should also keep `soma.io` separated, as it is not a core SOMA function, but a helper package.

**Comment by thetorpedodog** [comment]
> This makes sense to me. The only place I am waffling is storage-system-specific names, which I want to clearly disambiguate from other storage-system-specific names, for the day when we have multiple implementations. The alternative is that we could require these names to clearly disambiguate in their own name (eg, 'soma.SOMATileDBContext`vs`soma.tiledb.SOMAContext`). Either would work for me....

In this case, I think the namespace does the work for us: a `tiledbsoma.Context` is different from an `otherimpl.Context`.
 
> We should also keep `soma.io` separated, as it is not a core SOMA function, but a helper package.

Agreed; I was more referring to *core functionality* that we expect the user to call/instantiate, as opposed to helper that a user might use.

**Comment by bkmartinjr** [comment]
> In this case, I think the namespace does the work for us: a tiledbsoma.Context is different from an otherimpl.Context.

Right, that was my point - either use a sub-namespace, or clearly include storage engine name in the top-level name.

Example:    `soma.option.SOMATileDBContext` could be either `soma.tiledb.SOMAContext` or `soma.SOMATileDBContext`.

**Comment by johnkerl** [comment]
So then @bkmartinjr `soma.stats_enable()` -> `soma.tiledb_stats_enable()` -- ?

**Comment by johnkerl** [comment]
@bkmartinjr @thetorpedodog how's this for a draft for [the wiki](https://github.com/single-cell-data/TileDB-SOMA/wiki/Information-for-developers):

# Top-level namespacing

* Items in the top-level namespace should include things we expect a user to instantiate themselves or call directly
* Exception: helper packages such as `tiledbsoma.io`
* Anything storage-dependent should include the storage engine as a prefix or submodule: e.g. `soma.tiledb_foo()`
  * For Python: prefix or submodule
  * For: use a prefix since there are no submodules in R

**Comment by thetorpedodog** [comment]
> > In this case, I think the namespace does the work for us: a tiledbsoma.Context is different from an otherimpl.Context.
> 
> Right, that was my point - either use a sub-namespace, or clearly include storage engine name in the top-level name.
> 
> Example: `soma.option.SOMATileDBContext` could be either `soma.tiledb.SOMAContext` or `soma.SOMATileDBContext`.

I don’t follow what you’re saying here—when this is installed, it will be available at `tiledbsoma.whatever`; there is no `soma` namespace to add or change things in.

**Comment by johnkerl** [comment]
@thetorpedodog I think the (often-used) convention here is `import tiledbsoma as soma`

**Comment by bkmartinjr** [comment]
@thetorpedodog -- it is as John said.  We are trying to write mostly portable code, so we do:

> import tiledbsoma as soma

Even where it is non-portable, i.e., uses TileDB-specific names or features, we try to partition for clarity.  So ideally, I'd like the names to be clear when they are specific to a single implementation.

> So then soma.stats_enable() -> soma.tiledb_stats_enable() -- ?

As long as "stats_enable" is a TileDB-platform-specific feature, this would be my preference.   Or use a platform-specific namespace such as `soma.tiledb.stats_enable()`.  Either works fine for me, as they are clearly platform-specific in their naming, and won't be confused with some other `stats_enable` in storage-engine-portable code.

As an aside, we also have some code that can't be portable, but uses `get_storage_engine()` to figure out what it can do, eg.,

```
import tiledbsoma as soma

def consolidate_array(collection: soma.Collection):
    if collection.get_storage_engine() == 'tiledb':
        import tiledb
        tiledb.consolidate(...)
```

etc.

**Comment by johnkerl** [comment]
@bkmartinjr @thetorpedodog as there are no further objections I'll:

* Put this wording into this repo's wiki: https://github.com/single-cell-data/TileDB-SOMA/issues/915#issuecomment-1430183659
* Implement a PR for `stats_enable` -> `tiledb_stats_enable` (& the others _mutatis mutandis_) in R and Python both

**Comment by johnkerl** [comment]
#944 shows a problem: it fails R build since the TileDB-R package already has `tiledb::tiledb_stats_enable`

So we need `tiledbsoma_stats_enable` then.

**Comment by eddelbuettel** [comment]
Can you show a concrete quote?  I followed one of your links to another and to another.

It did build when I wrote it.  At least on one one OS.  Not that that proves much generality...

**Comment by johnkerl** [comment]
@eddelbuettel the error is on #944 and it's

```
RcppExports.cpp:106:6: error: ambiguating new declaration of ‘void tiledb_stats_enable()’
```

Yes, it was fine when you wrote it but it was not fine when I made on #944 the change suggested above which is `stats_*` -> `tiledb_stats_*`

**Comment by eddelbuettel** [comment]
Hm. I am pretty sure that wasn't produced when I committed what I commited.  Things may have changed since then.

**Comment by johnkerl** [comment]
@eddelbuettel no. Let me try again to be more clear.

* What you wrote -- what's in main -- has `stats_*`.
  * That is fine. What you did is fine. Except the discussion on this PR advocates for changing the naming convention to introduce a `tiledb_` prefix.
* What @bkmartinjr recommended in this issue was to change _from_ `stats_*` _to_ `tiledb_stats_*`.
  * For R, the change that I made on #944 _in order to implement the suggested naming convention `tiledb_stats_*` from our `tiledbsoma` now clashes with the existing names `tiledb_stats_*` from the `tiledb_` package.
* Possible solutions I can think of:
  * I update #944 to make the names be instead (for R and Python both) `tiledbsoma_stats_*`.
  * I update #944 to make the names be `tiledb_stats_*` for Python, and `tiledbsoma_stats_*` for R.
  * Personally I prefer the former.

**Comment by eddelbuettel** [comment]
Thanks for providing the fuller picture -- I now see the context more clearly.  And I did indeed choose `stats_` because `tiledb_stats_` was already taken.  I also prefer having both APIs use the same 'verbs' and names so `tiledbsoma_stats_`  may be best (even if it is a mouthful).

Have we considered not exporting it?  So `tiledbsoma.stats_` for Python and `tiledbsoma::stats_` for R?  If of course the design decision is to be a top-level accesor (and to be exported) then we likely have no choice but to fully prefix to provide the old-school way of a namespace layer.
